src/core/lib/security/credentials/external/url_external_account_credentials.cc

```diff
@@ -121,7 +123,18 @@ void UrlExternalAccountCredentials::RetrieveSubjectToken(
   grpc_httpcli_request request;
   memset(&request, 0, sizeof(grpc_httpcli_request));
   request.host = const_cast<char*>(url_.authority().c_str());
-  request.http.path = gpr_strdup(url_.path().c_str());
+  if (url_.query_parameter_pairs().empty()) {
+    request.http.path = gpr_strdup(url_.path().c_str());
+  } else {
+    std::vector<std::string> query_vector;
+    for (const URI::QueryParam& query_kv : url_.query_parameter_pairs()) {
+      query_vector.emplace_back(
+          absl::StrCat(query_kv.key, "=", query_kv.value));
+    }
+    std::string query = absl::StrJoin(query_vector, "&");
+    request.http.path =
+        gpr_strdup(absl::StrCat(url_.path(), "?", query).c_str());
```

Comment:
same here with url_.path(), though it was already a problem (albeit a much less common problem to run into).

Reply:
Should `URI` provide a method that returns the original path and query string? Because the `grpc_http_request` expects the original path and query strings, it would be a very common case to do `request.path = url.original_path()`. Otherwise, I gonna re-encode it, which reduce the efficiency a little bit.

Reply:
Storing the original URI string would approximately double the size of the URI object. I took a look, and given that this is the only place where re-encoding would be valuable at the moment, I'd recommend you add a way to percent-encode the query string here (at least), and not in the URI class. If it becomes a more general need, we can refactor things later.